src/core/instance/state.js

```diff
@@ -60,6 +60,16 @@ const isReservedProp = {
   slot: 1
 }
 
+function checkOptionType (vm: Component, name: string) {
+  const option = vm.$options[name]
+  if (!isPlainObject(option)) {
+    warn(
+      `${name} should be an object.`,
```

Comment:
The message could be a bit more explicit: `component option "${name}" should be an object`